onnxruntime/python/tools/transformers/gpt2_beamsearch_helper.py

```diff
@@ -0,0 +1,822 @@
+        )  # shape=(batch, beam_size^2)
+        output_log_probs, selected_index_flat = output_log_probs.topk(
+            self.config.beam_size, dim=-1, largest=True, sorted=True
+        )  # output shape=(batch, beam_size)
+
+        # select the correspondent sentences/next tokens
+        selected_input_seq = selected_index_flat // self.config.beam_size
+        next_token_ids = next_token_ids.view(self.config.batch_size, -1).gather(
+            -1, selected_index_flat
+        )
+
+        prev_step_results = prev_step_results.view(
+            self.config.batch_size, -1, prev_step_results.size(-1)
+        )
+        prev_step_results = prev_step_results.gather(
+            1, selected_input_seq.unsqueeze(-1).repeat(1, 1, prev_step_results.size(-1))
+        )
+
+        output_unfinished_sents = input_unfinished_sents.gather(1, selected_input_seq)
+        output_unfinished_sents = (
+            output_unfinished_sents
+            & next_token_ids.ne(self.config.eos_token_id)
+        )
+
+        # get the next full input_ids
+        current_step_results = torch.cat(
+            [prev_step_results, next_token_ids.unsqueeze(-1)], dim=-1
+        ).contiguous()
+
+        prev_step_scores = prev_step_scores.view(
+            self.config.batch_size, -1, prev_step_scores.size(-1)
+        )
+        prev_step_scores = prev_step_scores.gather(
+            1, selected_input_seq.unsqueeze(-1).repeat(1, 1, prev_step_scores.size(-1))
+        )
+        current_step_scores = torch.cat(
+            [prev_step_scores, output_log_probs.unsqueeze(-1)], dim=-1
+        ).contiguous()
+
+        return (
+            next_token_ids,
+            present_flat,
+            selected_input_seq,
+            output_log_probs,
+            output_unfinished_sents,
+            current_step_results.view(self.config.batch_size * self.config.beam_size, -1),
+            current_step_scores.view(self.config.batch_size * self.config.beam_size, -1),
+        )
+
+
+# Maps model class name to a tuple of model class, name of first output and use padding or not
+MODEL_CLASSES = {
+    'GPT2LMHeadModel': (MyGPT2LMHeadModel, 'logits', True),
+    'GPT2LMHeadModel_NoPadding': (MyGPT2LMHeadModel_NoPadding, 'logits', False),
+    'GPT2Model': (MyGPT2Model, 'last_state', True),
+    "GPT2LMHeadModel_BeamSearchStep": (GPT2LMHeadModel_BeamSearchStep, "last_state", True), # defined in gpt2_beamsearch_helper.py
+}
+
+
+class Gpt2BeamSearchInputs(Gpt2Inputs):
+    def __init__(
+        self,
+        input_ids,
+        position_ids,
+        attention_mask,
+        past,
+        beam_select_idx=None,
+        input_log_probs=None,
+        input_unfinished_sents=None,
+        prev_step_results=None,
+        prev_step_scores=None,
+    ):
+        super().__init__(input_ids, position_ids, attention_mask, past)
+        self.prev_step_results: torch.LongTensor = prev_step_results
+        self.prev_step_scores: torch.FloatTensor = prev_step_scores
```

Comment:
Union[torch.FloatTensor, torch.HalfTensor]? Is the input always in CPU?

When model is fp16, the data type could be torch.HalfTensor. 
If the input is in GPU, it could also be torch.cuda.FloatTensor.

Reply:
Updated. 

The input will always be in CPU.

Reply:
I have pushed a few changes. Now the input works on GPU as well.